File: bun-civitai-browser-master/src/civitai-api/v1/client/endpoints/model-versions.ts
import type { Result } from "neverthrow";
import { err, ok } from "neverthrow";
import type { CivitaiClient } from "../client";
import type { CivitaiError } from "../errors";
import type { ModelVersionEndpointData } from "../../models/model-version";

/**
 * ModelVersions endpoint interface
 */
export interface ModelVersionsEndpoint {
  /**
   * Get model version details
   */
  getById(id: number): Promise<Result<ModelVersionEndpointData, CivitaiError>>;

  /**
   * Get model version by hash
   */
  getByHash(
    hash: string,
  ): Promise<Result<ModelVersionEndpointData, CivitaiError>>;

  /**
   * Resolve file download URL to get the actual download URL
   * This method adds the token to the download URL and follows redirects
   * to get the final CDN URL for downloading the file
   */
  resolveFileDownloadUrl(
    fileDownloadUrl: string,
    token?: string,
  ): Promise<Result<string, CivitaiError>>;
}

/**
 * ModelVersions endpoint implementation
 */
export class ModelVersionsEndpointImpl implements ModelVersionsEndpoint {
  constructor(private readonly client: CivitaiClient) {}

  async getById(
    id: number,
  ): Promise<Result<ModelVersionEndpointData, CivitaiError>> {
    return this.client.get<ModelVersionEndpointData>(`model-versions/${id}`);
  }

  async getByHash(
    hash: string,
  ): Promise<Result<ModelVersionEndpointData, CivitaiError>> {
    return this.client.get<ModelVersionEndpointData>(
      `model-versions/by-hash/${hash}`,
    );
  }

  async resolveFileDownloadUrl(
    fileDownloadUrl: string,
    token?: string,
  ): Promise<Result<string, CivitaiError>> {
    // Use the client's download token if no token is provided
    const downloadToken =
      token ||
      this.client.getConfig().downloadToken ||
      this.client.getConfig().apiKey;

    if (!downloadToken) {
      // Create a simple error that matches CivitaiError structure
      const error: CivitaiError = {
        type: "UNAUTHORIZED",
        status: 401,
        message:
          "Download token is required to resolve file download URLs. Please provide apiKey or downloadToken in client configuration.",
        details: {
          suggestion:
            "Add your API key or download token to the client configuration",
        },
      };
      return err(error);
    }

    // Make a GET request with Authorization header to follow redirects and get the final URL
    // According to example.ts, we should use Authorization header, not token in URL
    try {
      const response = await fetch(fileDownloadUrl, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${downloadToken}`,
        },
        redirect: "follow", // Ensure we follow redirects to get the final CDN URL
      });

      if (!response.ok) {
        if (response.status === 401) {
          const error: CivitaiError = {
            type: "UNAUTHORIZED",
            status: 401,
            message: `Unauthorized to access model file download URL: ${fileDownloadUrl}. You may need to purchase the model on Civitai.`,
            details: {
              suggestion:
                "Check if you have access to this model or if your API key is valid",
            },
          };
          return err(error);
        } else {
          const error: CivitaiError = {
            type: "NETWORK_ERROR",
            status: response.status,
            code: `HTTP_${response.status}`,
            message: `Failed to resolve model file download URL: ${fileDownloadUrl}. Status: ${response.status} ${response.statusText}`,
            originalError: new Error(
              `HTTP ${response.status}: ${response.statusText}`,
            ),
          };
          return err(error);
        }
      }

      // Return the final URL after following redirects
      // response.url contains the final URL after all redirects
      return ok(response.url);
    } catch (error) {
      // Handle network errors
      const networkError: CivitaiError = {
        type: "NETWORK_ERROR",
        status: 0,
        code: "NETWORK_ERROR",
        message: `Network error while resolving download URL: ${error instanceof Error ? error.message : String(error)}`,
        originalError: error,
      };
      return err(networkError);
    }
  }
}

File: lib/civitai_api/endpoints/model_versions.dart
import 'package:dartz/dartz.dart';
import 'package:dio/dio.dart';

import '../errors.dart';
import '../http_client.dart';
import '../models/model_version.dart';

// ---------------------------------------------------------------------------
// Type
// ---------------------------------------------------------------------------

/// Model Versions API.
typedef ModelVersionsApi = ({
  Future<Either<CivitaiError, ModelVersionEndpointData>> Function(int) getById,
  Future<Either<CivitaiError, ModelVersionEndpointData>> Function(String)
  getByHash,
  Future<Either<CivitaiError, String>> Function(String, [String?])
  resolveFileDownloadUrl,
});

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/// Creates the [ModelVersionsApi] module, capturing [client] in closures.
///
/// [dio] is needed separately for `resolveFileDownloadUrl` because it follows
/// redirects to resolve CDN URLs — a raw Dio instance with redirect enabled
/// works better than the general [HttpClient] for this case.
ModelVersionsApi createModelVersionsApi(
  HttpClient client, {
  required Dio dio,
}) => (
  // GET /model-versions/:id
  getById: (int id) async {
    final result = await client.get('model-versions/$id');
    return result.flatMap(
      (json) => _decodeResponse<ModelVersionEndpointData>(
        json,
        ModelVersionEndpointData.fromJson,
        'model-version $id',
      ),
    );
  },

  // GET /model-versions/by-hash/:hash
  getByHash: (String hash) async {
    final result = await client.get('model-versions/by-hash/$hash');
    return result.flatMap(
      (json) => _decodeResponse<ModelVersionEndpointData>(
        json,
        ModelVersionEndpointData.fromJson,
        'model-version by hash $hash',
      ),
    );
  },

  // Resolve a file download URL → final CDN URL after redirects
  resolveFileDownloadUrl: (String fileUrl, [String? token]) async {
    // Use provided token, or get from Dio's auth header
    final authToken =
        token ??
        (dio.options.headers['Authorization'] as String?)?.replaceFirst(
          'Bearer ',
          '',
        );

    if (authToken == null || authToken.isEmpty) {
      return left(
        const CivitaiError.api(
          401,
          'Download token required. Provide apiKey in config.',
        ),
      );
    }

    try {
      // Use a separate Dio instance with redirect following enabled
      final resolveDio = Dio(
        BaseOptions(
          followRedirects: true,
          headers: {'Authorization': 'Bearer $authToken'},
        ),
      );
      final response = await resolveDio.getUri(Uri.parse(fileUrl));
      // response.realUri contains the final URL after redirects
      return right(response.realUri.toString());
    } on DioException catch (e) {
      final statusCode = e.response?.statusCode ?? 0;
      if (statusCode == 401) {
        return left(
          CivitaiError.api(
            statusCode,
            'Unauthorized to access file. You may need to purchase this model.',
          ),
        );
      }
      return left(
        CivitaiError.api(
          statusCode,
          'Failed to resolve download URL: ${e.message}',
        ),
      );
    } catch (e) {
      return left(CivitaiError.network('Network error resolving URL: $e', e));
    }
  },
);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

Either<CivitaiError, T> _decodeResponse<T>(
  dynamic json,
  T Function(Map<String, dynamic>) fromJson,
  String context,
) {
  if (json is! Map<String, dynamic>) {
    return left(CivitaiError.api(0, 'Unexpected response type for $context'));
  }
  try {
    return right(fromJson(json));
  } catch (e) {
    return left(CivitaiError.api(0, 'Failed to decode $context: $e'));
  }
}
